File: yogi-core/src/objects/configuration.cc
/*
 * This file is part of the Yogi Framework
 * https://github.com/yohummus/yogi-framework.
 *
 * Copyright (c) 2020 Johannes Bergmann.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <src/objects/configuration.h>
#include <src/objects/configuration/cmdline_parser.h>

#include <boost/algorithm/string.hpp>
#include <fstream>
#include <sstream>
#include <string>
using namespace std::string_literals;

YOGI_DEFINE_INTERNAL_LOGGER("Configuration")

namespace {

void check_circular_variable_dependency(const std::string& var_ref, const nlohmann::json& var_val) {
  if (var_val.is_string()) {
    auto str = var_val.get<std::string>();
    if (str.find(var_ref) != std::string::npos) {
      throw DescriptiveError{YOGI_ERR_UNDEFINED_VARIABLES} << "Circular dependency in variable \""
                                                           << var_ref.substr(2, var_ref.size() - 3) << '"';
    }
  }
}

template <typename Fn>
void walk_all_elements(nlohmann::json* json, Fn fn) {
  for (auto it = json->begin(); it != json->end(); ++it) {
    if (it.value().is_structured()) {
      walk_all_elements(&it.value(), fn);
    }

    if (json->is_object()) {
      fn(it.key(), &it.value());
    } else {
      static const std::string dummy;
      fn(dummy, &it.value());
    }
  }
}

void resolve_single_variable(nlohmann::json* elem, const std::string& var_ref, const nlohmann::json& var_val) {
  if (!elem->is_string()) {
    return;
  }

  auto val = elem->get<std::string>();
  if (val == var_ref) {
    *elem = var_val;
  } else {
    auto replacement = var_val.is_string() ? var_val.get<std::string>() : var_val.dump();
    boost::replace_all(val, var_ref, replacement);
    *elem = val;
  }
}

void resolve_variables_section(nlohmann::json* vars) {
  for (auto it = vars->begin(); it != vars->end(); ++it) {
    auto var_ref = "${"s + it.key() + '}';
    auto var_val = it.value();
    check_circular_variable_dependency(var_ref, var_val);

    for (auto& elem : *vars) {
      resolve_single_variable(&elem, var_ref, var_val);
    }
  }
}

nlohmann::json resolve_variables(const nlohmann::json& unresolved_json) {
  auto json = unresolved_json;
  if (!json.count("variables")) {
    return json;
  }

  auto& vars = json["variables"];

  resolve_variables_section(&vars);

  walk_all_elements(&json, [&](const auto&, auto* elem) {
    for (auto it = vars.cbegin(); it != vars.cend(); ++it) {
      auto var_ref = "${"s + it.key() + '}';
      resolve_single_variable(elem, var_ref, it.value());
    }
  });

  return json;
}

void check_variables_only_used_in_values(nlohmann::json* json) {
  walk_all_elements(json, [=](const auto& key, auto) {
    if (key.find("${") != std::string::npos) {
      throw DescriptiveError{YOGI_ERR_VARIABLE_USED_IN_KEY} << "Found syntax for variable in key: " << key;
    }
  });
}

void check_all_variables_are_resolved(nlohmann::json* json) {
  walk_all_elements(json, [=](const auto&, const auto* elem) {
    if (!elem->is_string()) return;
    auto val = elem->template get<std::string>();
    auto pos = val.find("${");
    if (pos != std::string::npos) {
      throw DescriptiveError{YOGI_ERR_UNDEFINED_VARIABLES}
          << "Variable \"" << val.substr(pos + 2, val.find('}', pos + 2) - pos - 2) << "\" could not be resolved";
    }
  });
}

}  // anonymous namespace

Configuration::Configuration(int cfg_flags)
    : variables_supported_{!(cfg_flags & YOGI_CFG_DISABLE_VARIABLES)},
      mutable_cmdline_{!!(cfg_flags & YOGI_CFG_MUTABLE_CMD_LINE)},
      json_({}),
      immutable_json_({}) {
  set_logging_prefix(*this);
}

void Configuration::update_from_cmdline(int argc, const char* const* argv, int clo_flags) {
  CmdlineParser parser(argc, argv, clo_flags);
  parser.parse();

  verify_and_merge(parser.get_files_configuration(), parser.get_direct_configuration());

  if (!mutable_cmdline_) {
    immutable_json_ = parser.get_direct_configuration();
  }
}

void Configuration::update_from_string(const char* json_str) {
  nlohmann::json json;

  try {
    json = nlohmann::json::parse(json_str);
  } catch (const nlohmann::json::exception& e) {
    throw DescriptiveError{YOGI_ERR_PARSING_JSON_FAILED} << "Could not parse JSON string: " << e.what();
  }

  verify_and_merge(json, immutable_json_);
}

void Configuration::update_from_file(const char* filename) {
  std::ifstream f(filename);
  if (!f.is_open() || f.fail()) {
    throw DescriptiveError{YOGI_ERR_PARSING_FILE_FAILED} << "Could not open " << filename;
  }

  nlohmann::json json;
  try {
    f >> json;
  } catch (const std::exception& e) {
    throw DescriptiveError{YOGI_ERR_PARSING_FILE_FAILED} << "Could not parse " << filename << ": " << e.what();
  }

  verify_and_merge(json, immutable_json_);
}

nlohmann::json Configuration::get_json(bool resolve_vars, const char* section) const {
  nlohmann::json json;
  nlohmann::json::json_pointer jp;

  if (section != nullptr) {
    try {
      jp = nlohmann::json::json_pointer(section);
    } catch (const nlohmann::json::exception& e) {
      throw DescriptiveError{YOGI_ERR_INVALID_PARAM} << "Could not parse JSON pointer: " << e.what();
    }
  }

  if (resolve_vars) {
    if (!variables_supported_) {
      throw Error{YOGI_ERR_NO_VARIABLE_SUPPORT};
    }

    json = resolve_variables(json_)[jp];
  } else {
    json = json_[jp];
  }

  if (!json.is_object()) {
    throw DescriptiveError{YOGI_ERR_CONFIGURATION_SECTION_NOT_FOUND} << "Could not find section \"" << section
                                                                     << "\" in the configuration.";
  }

  return json;
}

nlohmann::json Configuration::get_json(const char* section) const { return get_json(variables_supported_, section); }

std::string Configuration::dump(bool resolve_variables, int indentation_width) const {
  return get_json(resolve_variables).dump(indentation_width);
}

void Configuration::write_to_file(const char* filename, bool resolve_vars, int indentation_width) const {
  if (!variables_supported_ && resolve_vars) {
    throw Error{YOGI_ERR_NO_VARIABLE_SUPPORT};
  }

  std::ofstream f(filename);
  if (!f.is_open() || f.fail()) {
    throw Error{YOGI_ERR_READ_FILE_FAILED};
  }

  try {
    if (resolve_vars) {
      f << resolve_variables(json_).dump(indentation_width);
    } else {
      f << json_.dump(indentation_width);
    }

    if (indentation_width != -1) {
      f << std::endl;
    }
  } catch (const std::exception& e) {
    LOG_ERR("Could not write configuration to " << filename << ": " << e.what());
    throw Error{YOGI_ERR_WRITE_FILE_FAILED};
  }
}

void Configuration::verify_and_merge(const nlohmann::json& json_to_merge, const nlohmann::json& immutable_json) {
  auto new_json = json_;
  new_json.merge_patch(json_to_merge);
  new_json.merge_patch(immutable_json);

  if (variables_supported_) {
    check_variables_only_used_in_values(&new_json);
    auto resolved_json = resolve_variables(new_json);
    check_all_variables_are_resolved(&resolved_json);
  }

  json_ = new_json;
}

File: yogi-dotnet/yogi/Configuration.cs
/*
 * This file is part of the Yogi Framework
 * https://github.com/yohummus/yogi-framework.
 *
 * Copyright (c) 2020 Johannes Bergmann.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

using System;
using System.Runtime.InteropServices;
using System.Collections.Generic;
using Newtonsoft.Json.Linq;

public static partial class Yogi
{
    /// <summary>
    /// A configuration represents a set of parameters that usually remain constant
    /// throughout the runtime of a program. Parameters can come from different
    /// sources such as the command line or a file. Configurations are used for
    /// other parts of the library such as application objects, however, they are
    /// also intended to store user-defined parameters.
    /// </summary>
    public class Configuration : Object
    {
        /// <summary>
        /// Constructor.
        /// </summary>
        /// <param name="flags">Flags for behaviour adjustments.</param>
        public Configuration(ConfigurationFlags flags = ConfigurationFlags.None)
        : base(Create(flags))
        {
            Flags = flags;
        }

        /// <summary>
        /// Updates the configuration from command line options.
        ///
        /// If parsing the command line, files or any given JSON string fails, or
        /// if help is requested (e.g. by using the --help switch) then a
        /// DetailedFailureException will be raised containing detailed
        /// information about the error or the help text.
        /// </summary>
        /// <param name="argv">List of command line arguments including the script name.</param>
        /// <param name="options">Options to provide on the command line.</param>
        public void UpdateFromCommandLine(IEnumerable<string> argv,
                                          CommandLineOptions options = CommandLineOptions.None)
        {
            var args = new List<string>();
            args.AddRange(argv);

            int res = YogiCore.YOGI_ConfigurationUpdateFromCommandLine(Handle, args.Count, args.ToArray(),
                                                                       (int)options);
            CheckErrorCode(res);
        }

        /// <summary>
        /// Updates the configuration from a JSON object serialized to a string.
        ///
        /// If parsing fails then a DetailedFailureException with the
        /// ConfigurationValidationFailed error will be raised containing detailed
        /// information about the failure.
        /// </summary>
        /// <param name="json">JSON object.</param>
        public void UpdateFromJson(JsonView json)
        {
            int res = YogiCore.YOGI_ConfigurationUpdateFromJson(Handle, json.Data);
            CheckErrorCode(res);
        }

        /// <summary>
        /// Updates the configuration from a JSON file.
        ///
        /// If parsing fails then a DetailedFailureException with the
        /// ConfigurationValidationFailed error will be raised containing detailed
        /// information about the failure.
        /// </summary>
        /// <param name="filename">Path to the JSON file.</param>
        public void UpdateFromFile(string filename)
        {
            int res = YogiCore.YOGI_ConfigurationUpdateFromFile(Handle, filename);
            CheckErrorCode(res);
        }

        /// <summary>
        /// Retrieves the configuration as a JSON-formatted string.
        /// </summary>
        /// <param name="resolveVariables">Resolve all configuration variables.
        /// By default, variables get resolved if the configuration supports them.</param>
        /// <param name="indentation">Number of space characters to use for indentation.
        /// By default, no spaces are used and new lines are omitted as well.</param>
        /// <returns>The configuration serialized to a string.</returns>
        public string Dump([Optional] bool? resolveVariables, [Optional] int? indentation)
        {
            if (resolveVariables == null)
            {
                resolveVariables = (Flags & ConfigurationFlags.DisableVariables)
                                    != ConfigurationFlags.DisableVariables;
            }

            if (indentation == null)
            {
                indentation = -1;
            }

            var json = new IntPtr();
            int jsonsize = 0;
            int res = YogiCore.YOGI_ConfigurationDump(Handle, ref json, ref jsonsize, resolveVariables == true ? 1 : 0,
                                                      (int)indentation);
            CheckErrorCode(res);
            return Marshal.PtrToStringAnsi(json);
        }

        /// <summary>
        /// Retrieves the configuration as a JSON object.
        /// </summary>
        /// <param name="resolveVariables">Resolve all configuration variables.
        /// By default, variables get resolved if the configuration supports them.</param>
        /// <returns>JSON object representing the configuration.</returns>
        public JObject ToJson([Optional] bool? resolveVariables)
        {
            return JObject.Parse(Dump(resolveVariables));
        }

        /// <summary>
        /// Writes the configuration to a file in JSON format.
        /// </summary>
        /// <param name="filename">Path to the output file.</param>
        /// <param name="resolveVariables">Resolve all configuration variables.
        /// By default, variables get resolved if the configuration supports them.</param>
        /// <param name="indentation">Number of space characters to use for indentation.
        /// By default, no spaces are used and new lines are omitted as well.</param>
        public void WriteToFile(string filename, [Optional] bool? resolveVariables,
            [Optional] int? indentation)
        {
            if (resolveVariables == null)
            {
                resolveVariables = (Flags & ConfigurationFlags.DisableVariables)
                                    != ConfigurationFlags.DisableVariables;
            }

            if (indentation == null)
            {
                indentation = -1;
            }

            int res = YogiCore.YOGI_ConfigurationWriteToFile(Handle, filename, resolveVariables == true ? 1 : 0,
                                                             (int)indentation);
            CheckErrorCode(res);
        }

        /// <summary>
        /// Validates a section of the configuration against a JSON Schema.
        ///
        /// The validation is based on JSON Schema draft-07, see
        /// http://json-schema.org/. The schema to validate against has to be
        /// supplied in \p schema which needs to be a configuration object itself.
        ///
        /// If the validation fails, the kConfigurationValidationFailed error will be
        /// thrown, containing a human-readable description about the failure.
        /// </summary>
        /// <param name="section">Section in the configuration to validate; syntax is
        /// JSON pointer (RFC 6901).</param>
        /// <param name="schema">The schema to use.</param>
        public void Validate(string section, Configuration schema)
        {
            int res = YogiCore.YOGI_ConfigurationValidate(Handle, section, schema.Handle);
            CheckErrorCode(res);
        }

        /// <summary>
        /// Validates the configuration against a JSON Schema.
        ///
        /// The validation is based on JSON Schema draft-07, see
        /// http://json-schema.org/. The schema to validate against has to be
        /// supplied in \p schema which needs to be a configuration object itself.
        ///
        /// If the validation fails, the kConfigurationValidationFailed error will be
        /// thrown, containing a human-readable description about the failure.
        /// </summary>
        /// <param name="section">Section in the configuration to validate; syntax is
        /// JSON pointer (RFC 6901).</param>
        /// <param name="schema">The schema to use.</param>
        public void Validate(Configuration schema)
        {
            Validate(null, schema);
        }

        /// <summary>
        /// Configuration flags.
        /// </summary>
        public readonly ConfigurationFlags Flags;

        /// <summary>
        /// Serializes the configuration to a string.
        /// </summary>
        /// <returns>Configuration serialized to a string.</returns>
        public override string ToString()
        {
            return Dump();
        }

        static IntPtr Create(ConfigurationFlags flags)
        {
            var handle = new IntPtr();
            int res = YogiCore.YOGI_ConfigurationCreate(ref handle, (int)flags);
            CheckErrorCode(res);
            return handle;
        }
    }
}
